File: frontend/src/types/promotion.ts
// Promotion Types (백엔드 PromotionResponse 스펙 기준)
export interface Promotion {
  // 백엔드 PromotionResponse 필드
  id: number;
  company?: string;
  productName: string;
  originalPrice: number;
  salePrice: number;
  discountRate: number;
  totalQuantity: number;
  startTime: string;
  endTime: string;
  productImageUrl?: string;
  
  // 프론트엔드 호환성을 위한 별칭 (optional)
  productImage?: string;     // productImageUrl의 별칭
  
  // 백엔드에서 추가로 제공할 수 있는 필드 (optional)
  adminId?: number;
  productId?: number;
  promotionStatus?: PromotionStatus;
  soldQuantity?: number;
}

// 백엔드 Enum: SCHEDULER, ACTIVE, ENDED
export type PromotionStatus = 'SCHEDULER' | 'ACTIVE' | 'ENDED';

// 백엔드 PromotionListResponse (상태별 조회 응답)
export interface PromotionListResponse {
  promotions: Promotion[];
}

// 백엔드 PromotionRequest (POST/PUT)
export interface PromotionRequest {
  adminId?: number;
  productId: number;
  discountRate: number;
  startTime: string;        // LocalDateTime
  endTime: string;          // LocalDateTime
  totalQuantity: number;
}

File: src/main/java/com/ohgiraffers/timedeal/core/domain/Promotion.java
package com.ohgiraffers.timedeal.core.domain;

import com.ohgiraffers.timedeal.core.enums.EntityStatus;
import com.ohgiraffers.timedeal.core.enums.PromotionStatus;
import com.ohgiraffers.timedeal.core.support.error.CoreException;
import com.ohgiraffers.timedeal.core.support.error.ErrorType;
import com.ohgiraffers.timedeal.storage.BaseEntity;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "promotions")
public class Promotion extends BaseEntity {

    @Column(name = "admin_id" ,  nullable = false)
    private Long adminId;

    @Column(name = "product_id", nullable = false)
    private Long productId;

    @Column(name = "sale_price")
    private Double salePrice;

    @Column(name = "discount_rate")
    private Double discountRate;

    @Column(name = "start_time")
    private LocalDateTime startTime;

    @Column(name = "end_time")
    private LocalDateTime endTime;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "total_quantity")
    private Integer totalQuantity;

    @Column(name = "sold_quantity")
    private Integer soldQuantity;

    @Column(name = "promotion_status")
    @Enumerated(EnumType.STRING)
    private PromotionStatus promotionStatus = PromotionStatus.ACTIVE;

    public void increaseSoldQuantity() {
        if(this.soldQuantity > this.totalQuantity) {
            throw new CoreException(ErrorType.DEFAULT_ERROR);
        }
        this.soldQuantity += 1;
    }
}
